multidim_threshold/projection.py

```diff
@@ -35,26 +36,37 @@ def generate_boundary_approxes(lo, hi, member_oracles, **kwargs):
         yield boundaries
 
 
-def generate_projections(lo, hi, member_oracles, *, direc=None, searches=None):
-    proj_vecs = generate_proj_vecs(lo, hi, direc)
+def generate_projections(lo, hi, member_oracles, *, direc=None, searches=None, random=False):
+    axis_hi = axes_intersects(lo, hi, member_oracles, random=False) if random else hi
+
+    proj_vecs = generate_proj_vecs(lo, axis_hi, direc, random)
+
     if searches is None:
         searches = [learn_search(f, lo) for f in member_oracles]
 
     for vec in proj_vecs:
         yield projections(hi, vec, searches=searches)
 
 
-def generate_proj_vecs(lo, hi, direc=None):
+def generate_proj_vecs(lo, hi, direc=None, random=False):
     lo, hi = mdt.map_array((lo, hi))
     if direc is None:
         direc = hi - lo
 
     vecs = [ProjVec(lo, direc)]
     while True:
         yield from vecs
-        vecs = [ProjVec(r, direc) for r in next_roots(lo, hi, vecs)]
+        vecs = [ProjVec(random_root(lo, axis_hi), direc) if random else ProjVec(r, direc) for r in next_roots(lo, hi, vecs)]
 
 
+def axes_intersects(lo, hi, member_oracles, *, searches=None):
+    if searches is None:
```

Comment:
probably better to just pass searches in always. This is just a helper function used once, so no need to make it super convenient api for it.